stata_kernel/kernel.py

```diff
@@ -116,13 +126,82 @@ def do_execute(
 
         """
 
-        cm = CodeManager(re.sub(r'\r\n', r'\n', code))
+        #############
+        #  Magics!  #
+        #############
+
+        magic, code = poor_mans_magic(code)
+        quit_early  = {'execution_count': self.execution_count}
+        quit_early['status'] = 'ok'
+        quit_early['payload'] = []
+        quit_early['user_expressions'] = {}
+
+        if magic['name'] in stata_magics['global_magics']:
+            obj      = self.do("macro dir")
+            res      = obj.get('res').rstrip()
+            sglobals = magic['regex'].findall(res)
+            lens     = 0
+            pglobals = []
+            if len(sglobals) > 0:
+                for macro, cr, contents in sglobals:
+                    if magic['locals'] and not macro.startswith('_'):
+                        continue
+
+                    if not magic['locals'] and macro.startswith('_'):
+                        continue
+
+                    if macro.startswith('_'):
+                        macro = macro[1:]
+                        extra = 1
+                    else:
+                        extra = 0
+
+                    if magic['gregex'] != '':
+                        if not magic['gregex'].search(macro):
+                            continue
+
+                    macro  += ':'
+                    lmacro  = len(macro)
+                    lspaces = len(cr.strip('\r\n'))
+                    lens    = max(lens, lmacro)
+                    if len(macro) <= 15:
+                        if (lspaces + lmacro + extra) > 16:
+                            pglobals += ((macro, ' ' + contents),)
+                        else:
+                            pglobals += ((macro, contents),)
+                    else:
+                        pglobals += ((macro, contents.lstrip('\r\n')),)
+
+            fmt = "{{0:{0}}} {{1}}".format(lens)
+            for macro, contents in pglobals:
+                print(fmt.format(macro, magic['sregex'].sub((lens + 1) * ' ', contents)))
+
+            return quit_early
+
+        if magic['status'] == -1:
+            return quit_early
+
+        ##########
+        #  Code  #
+        ##########
+
+        cm   = CodeManager(re.sub(r'\r\n', r'\n', code))
         code = cm.remove_comments()
+
         graph_keywords = [
-            r'gr(a|ap|aph)?', r'tw(o|ow|owa|oway)?',
-            r'sc(a|at|att|atte|atter)?', r'line']
+            r'gr(a|ap|aph)?',
+            r'tw(o|ow|owa|oway)?',
+            r'sc(a|at|att|atte|atter)?',
+            r'hist(ogram)?',
+            r'line',
+            r'kdensity',
+            r'lowess',
+            r'lpoly',
+            r'tsline'
+        ]
```

Comment:
I discovered that there's actually a _lot_ of graph keywords. See `help graph_other`.

A simple list is fine but eventually I'll want to at least copy and paste the other graph keywords into the list.